userd/autostart.go

```diff
@@ -0,0 +1,210 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package userd
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil/shlex"
+)
+
+// expandDesktopFields processes the input string and expands any %<char>
+// patterns. '%%' expands to '%', all other patterns expand to empty strings.
+func expandDesktopFields(in string) string {
+	raw := []rune(in)
+	out := make([]rune, 0, len(raw))
+
+	var hasKey bool
+	for _, r := range raw {
+		if hasKey {
+			hasKey = false
+			// only allow %% -> % expansion, drop other keys
+			if r == '%' {
+				out = append(out, r)
+			}
+			continue
+		} else if r == '%' {
+			hasKey = true
+			continue
+		}
+		out = append(out, r)
+	}
+	return string(out)
+}
+
+func findExec(desktopFileContent []byte) (string, error) {
+	scanner := bufio.NewScanner(bytes.NewBuffer(desktopFileContent))
+	execCmd := ""
+	for scanner.Scan() {
+		bline := scanner.Bytes()
+
+		if !bytes.HasPrefix(bline, []byte("Exec=")) {
+			continue
+		}
+
+		full := string(bline[len("Exec="):])
+		execCmd = expandDesktopFields(full)
+		break
+	}
+
+	execCmd = strings.TrimSpace(execCmd)
+	if execCmd == "" {
+		return "", fmt.Errorf("Exec not found or invalid")
+	}
+	return execCmd, nil
+}
+
+func getCurrentSnapInfo(snapName string) (*snap.Info, error) {
+	curFn := filepath.Join(dirs.SnapMountDir, snapName, "current")
+	realFn, err := os.Readlink(curFn)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find current revision for snap %s: %s", snapName, err)
+	}
+	rev := filepath.Base(realFn)
+	revision, err := snap.ParseRevision(rev)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read revision %s: %s", rev, err)
+	}
+
+	return snap.ReadInfo(snapName, &snap.SideInfo{Revision: revision})
+}
+
+func autostartCmd(snapName, desktopFilePath string) (*exec.Cmd, error) {
+	desktopFile := filepath.Base(desktopFilePath)
+
+	info, err := getCurrentSnapInfo(snapName)
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain snap information for snap %q: %v", snapName, err)
```

Comment:
how about `cannot get command for starting snap %q...` since the error about the snap information is already in err?

Reply:
Hmm, passing the error as returned by `snap.ReadCurrentInfo()` should be fine as it will already contain enough context (eg. `cannot fund current revision for snap foo..` or `cannot read revision ..`).